nmxact/xact/image.go

```diff
@@ -185,33 +204,149 @@ func nextImageUploadReq(s sesn.Sesn, upgrade bool, data []byte, off int, imageNu
+func (t *ImageUploadIntTracker) update(t_off int, r_off int, status int) {
+	if status == IMAGE_UPLOAD_STATUS_MISSED {
+		t.rsp_map[t_off] = IMAGE_UPLOAD_CHUNK_MISSED_WM
+		return
+	}
 
-	for off := c.StartOff; off < len(c.Data); {
-		r, err := nextImageUploadReq(s, c.Upgrade, c.Data, off, c.ImageNum)
-		if err != nil {
-			return nil, err
-		}
+	if _, ok := t.rsp_map[t_off]; ok {
+		// send chunk@offset, increase count
+		t.rsp_map[t_off] += 1
+	} else {
+		// send chunk@offset for the first time
+		t.rsp_map[t_off] = 1
+	}
 
-		rsp, err := txReq(s, r.Msg(), &c.CmdBase)
-		if err != nil {
-			return nil, err
+	if _, ok := t.rsp_map[r_off]; ok {
+		// chunk@offset requested, in transit
+		t.rsp_map[r_off] -= 1
+	} else {
+		// chunk@offset was not sent
+		t.rsp_map[r_off] = -1
+	}
+
+}
+
+func (t *ImageUploadIntTracker) check_window(end int) bool {
+	t.mutex.Lock()
+	ret := false
+	defer t.mutex.Unlock()
+	if t.w_count < t.w_cap && t.off < end {
+		ret = true
+	}
+	return ret
+}
+
+func (t *ImageUploadIntTracker) process_missed_chunks() {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	// increase the current window count
+	t.w_count++
+	for o, c := range t.rsp_map {
+		if c < -(IMAGE_UPLOAD_MAX_WS * 2) {
+			delete(t.rsp_map, o)
+			t.off = o
+			log.Debugf("missed? off %d count %d", o, c)
 		}
-		irsp := rsp.(*nmp.ImageUploadRsp)
+	}
+}
 
-		off = int(irsp.Off)
+func (t *ImageUploadIntTracker) handle_response(c *ImageUploadCmd, off int, rsp nmp.NmpRsp, res *ImageUploadResult) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	irsp := rsp.(*nmp.ImageUploadRsp)
+	res.Rsps = append(res.Rsps, irsp)
+	t.update(off, int(irsp.Off), 0)
+	if c.ProgressCb != nil {
+		c.ProgressCb(c, irsp)
+	}
+	if t.tune_ws && t.w_cap < IMAGE_UPLOAD_MAX_WS {
+		t.w_cap++
+	}
+	t.w_count--
+}
 
-		if c.ProgressCb != nil {
-			c.ProgressCb(c, irsp)
+func (t *ImageUploadIntTracker) handle_error(off int, err error) {
+	/*XXX: there could be an Unauthorize or EOF error  when the rate is too high
+	  due to a large window example:
+	  "failed to decrypt message: coap_sec_tunnel: decode GCM fail EOF"
+	  Since the error is sent with fmt.Errorf() API, with no code,
+	  the string may have to be parsed to know the particular error */
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	if t.w_count > IMAGE_UPLOAD_START_WS+1 {
+		t.w_cap--
+	}
+	t.tune_ws = false
+	t.w_count--
+	t.update(off, -1, IMAGE_UPLOAD_STATUS_MISSED)
+}
+
+func (c *ImageUploadCmd) Run(s sesn.Sesn) (Result, error) {
+	res := newImageUploadResult()
+	var m sync.Mutex
+	sem := sync.NewCond(&m)
+	rsp_c := make(chan nmp.NmpRsp, IMAGE_UPLOAD_MAX_WS)
+	err_c := make(chan error, IMAGE_UPLOAD_MAX_WS)
+
+	t := ImageUploadIntTracker{
+		tune_ws: true,
+		w_count: 0,
+		w_cap:   IMAGE_UPLOAD_START_WS,
+		off:     c.StartOff,
+		rsp_map: make(map[int]int),
+	}
+
+	var wg sync.WaitGroup
+
+	for t.off < len(c.Data) {
+		// Skip the first offset as there is nothing to wait for
+		if t.off != 0 {
+			m.Lock()
+			sem.Wait()
+			m.Unlock()
```

Comment:
This makes me a bit uneasy.  I totally believe it works in practice.  In theory, there is no guarantee that we'll get around to waiting on this condition variable before all the goroutines signal it.

Alternative idea: use an atomic int that keeps track of the amount of unused window capacity.  If the capacity is 0, this code would block on a channel.  This is more appealing to me because channels remember that they've been signaled even if no one is listening at the time.  It would also allow you to get rid of that mutex.